File: backend/app/models/events.py
from typing import Literal

from pydantic import BaseModel

from app.models.metrics import FaceData, TextData, VoiceData


class UserMessage(BaseModel):
    text: str
    input_mode: Literal["voice", "text"]


class SelfReportScore(BaseModel):
    timing: Literal["pre_session", "post_session"]
    score: float  # 0.0 - 1.0


class FaceEvent(BaseModel):
    type: Literal["face"] = "face"
    data: FaceData


class VoiceEvent(BaseModel):
    type: Literal["voice"] = "voice"
    data: VoiceData


class TextEvent(BaseModel):
    type: Literal["text"] = "text"
    data: TextData


class MessageEvent(BaseModel):
    type: Literal["message"] = "message"
    data: UserMessage

File: frontend/src/types/events.ts
// WebSocket イベント型定義
// バックエンドの app/routers/websocket.py のプロトコルに対応

// ── クライアント → サーバー ───────────────────────────────────────────────────

export interface FaceEvent {
  type: "face";
  data: {
    happy: number;
    angry: number;
    sad: number;
    neutral: number;
    stability: number;
  };
}

export interface VoiceEvent {
  type: "voice";
  data: string; // Base64 WAV
}

export interface TextEvent {
  type: "text";
  data: {
    interval_ms: number;
    backspace_count: number;
    total_keys: number;
    idle_ms: number;
    total_time_ms: number;
  };
}

export interface SelfReportEvent {
  type: "self_report";
  timing: "pre_session" | "post_session";
  score: number; // 0.0 - 1.0
}

export interface MessageEvent {
  type: "message";
  data: {
    text: string;
    input_mode: "voice" | "text";
  };
}

export type ClientEvent = FaceEvent | VoiceEvent | TextEvent | SelfReportEvent | MessageEvent;

// ── サーバー → クライアント ───────────────────────────────────────────────────

export interface AckEvent {
  type: "ack";
  timing: "pre_session" | "post_session";
}

export interface StateEvent {
  type: "state";
  mental_state: string; // "A" | "B" | "C" | "D-1" | "D-2"
  score: number;
  reasoning: string;
  score_breakdown: Record<string, unknown>;
}

export interface QuestionEvent {
  type: "question";
  text: string;
}

export interface CloseEvent {
  type: "close";
}

export interface ErrorEvent {
  type: "error";
  message: string;
}

export type ServerEvent = AckEvent | StateEvent | QuestionEvent | CloseEvent | ErrorEvent;
